core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java

```diff
@@ -0,0 +1,99 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.sql.planner;
+
+import com.google.common.collect.ImmutableMap;
+import io.trino.Session;
+import io.trino.metadata.Metadata;
+import io.trino.operator.scalar.ArrayDistinctFunction;
+import io.trino.operator.scalar.ArraySortFunction;
+import io.trino.spi.type.Type;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.ExpressionRewriter;
+import io.trino.sql.tree.ExpressionTreeRewriter;
+import io.trino.sql.tree.FunctionCall;
+import io.trino.sql.tree.NodeRef;
+import io.trino.sql.tree.QualifiedName;
+import io.trino.sql.tree.SymbolReference;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public final class ArraySortAfterArrayDistinctRewriter
+{
+    public static Expression rewrite(Expression expression, Map<NodeRef<Expression>, Type> expressionTypes, Metadata metadata)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(expressionTypes, metadata), expression);
+    }
+
+    private ArraySortAfterArrayDistinctRewriter() {}
+
+    public static Expression rewrite(Expression expression, Session session, Metadata metadata, TypeAnalyzer typeAnalyzer, TypeProvider typeProvider)
+    {
+        requireNonNull(metadata, "metadata is null");
+        requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+
+        if (expression instanceof SymbolReference) {
+            return expression;
+        }
+        Map<NodeRef<Expression>, Type> expressionTypes = typeAnalyzer.getTypes(session, typeProvider, expression);
+
+        return rewrite(expression, expressionTypes, metadata);
+    }
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        private final Map<NodeRef<Expression>, Type> expressionTypes;
+        private final Metadata metadata;
+
+        public Visitor(Map<NodeRef<Expression>, Type> expressionTypes, Metadata metadata)
+        {
+            this.expressionTypes = ImmutableMap.copyOf(requireNonNull(expressionTypes, "expressionTypes is null"));
+            this.metadata = metadata;
+        }
+
+        @Override
+        public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (metadata.getFunctionMetadata(metadata.decodeFunction(node.getName())).getActualName().equals(ArrayDistinctFunction.NAME)
+                    && node.getArguments().size() > 0
+                    && node.getArguments().get(0) instanceof FunctionCall
+                    && metadata.getFunctionMetadata(metadata.decodeFunction(((FunctionCall) node.getArguments().get(0)).getName())).getActualName().equals(ArraySortFunction.NAME)) {
+                FunctionCall sortFunctionCall = (FunctionCall) node.getArguments().get(0);
```

Comment:
Call `treeRewriter.defaultRewrite()` on the `sortFunctionCall` to process the arguments recursively. Add test with nested rewrite.